### PR TITLE
feat(post-scan action): GCP Teams notification for FSS

### PR DESCRIPTION
**Pull request by GeorgeDavis-TM**
# GCP Teams notification for FSS

## Change Summary
- Added code and serverless framework deployment method for MS Teams notifications based on FSS detections on GCP
- Updated the code based on previous PR requests
- Switched from Serverless stage variables to params

## PR Checklist

- [x] I've read and followed the [Contributing Guide](https://github.com/trendmicro/cloudone-filestorage-plugins/blob/master/.github/CONTRIBUTING.md).
- Documents/Readmes
  - [x] Updated accordingly
- Plugins that have versioning
  - [x] Version bumped and follows [Semantic Versioning](https://semver.org/)

## Other Notes

- ~~⚠️ Recommend to merge PR #87 before merging this PR.~~

### CODE REVIEW

**Comment by GeorgeDavis-TM** [comment]
@trendmicro/cloud-one-file-storage-security Appreciate your time to review this new plugin. Looking forward to working with you and closing this PR 🎉

File: post-scan-actions/gcp-python-teams-notification/main.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        print(f'Event: {event}')
+        base64_data = event.get('data', '')
+        base64_bytes = base64_data.encode('ascii')
+        message_bytes = base64.b64decode(base64_bytes)
+        message = json.loads(message_bytes.decode('ascii'))
+        print(f'Message: {message}')
+
+        url = os.environ['TEAMS_URL']
+
+        print("""This Function was triggered by messageId {} published at {} to {}""".format(context.event_id, context.timestamp, context.resource))
```
Comment (luhsuanchen) [inline post-scan-actions/gcp-python-teams-notification/main.py:35]: What are these `print` used for? If they are used for logging, I suggest use old style formatting according to [logging module doc](https://docs.python.org/3/howto/logging.html#optimization) ([Django doctor](https://blog.devgenius.io/when-python-old-style-string-formatting-is-best-practice-691a0759b37e) also provides a thorough explanation).
While they are used for merely printing messages, I suggest all `print` use f-string instead to make your code have a unified style.

Comment (GeorgeDavis-TM) [reply post-scan-actions/gcp-python-teams-notification/main.py:35]: @luhsuanchen Appreciate your feedback. I intended to change everything to use the `logging` package but got lost in other FSS plugin dev. I will change it to `logging` with a commit to the branch right away.

**Comment by GeorgeDavis-TM** [comment]
:warning: Recommend to merge PR #87 before merging this PR.

**Comment by GeorgeDavis-TM** [comment]
@trendmicro/cloud-one-file-storage-security LGTM!

**Comment by GeorgeDavis-TM** [comment]
@trendmicro/cloud-one-file-storage-security Do we have an update on this PR? Please advise.

**Comment by GeorgeDavis-TM** [comment]
@andrew-c-lee Could you please review this PR as well when you have the time? I have made changes to this PR as and when I receive recommendations for PR #87. 

Thank you in advance.

**Comment by andrew-c-lee** [comment]
Hi @davidblanchard123, please help review README.md and docs/deploy-tutorial.md. Thanks

File: post-scan-actions/gcp-python-teams-notification/README.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
+# Cloud One File Storage Security Post Scan Action for GCP - MS Teams notifications
+
+**:warning: Note: File Storage Security for GCP solution is in preview.**
+
+## Prerequisites
+
+1. **Install supporting tools**
```
Comment (davidblanchard123) [inline post-scan-actions/gcp-python-teams-notification/README.md:7]:
```suggestion
1. Install Google Cloud SDK.
```

File: post-scan-actions/gcp-python-teams-notification/README.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+   - [Google Cloud SDK](https://cloud.google.com/sdk/docs/install-sdk)
+
+## Installation
+
+### With GCP Cloud Shell
+
+1. Open in Cloud Shell
```
Comment (davidblanchard123) [inline post-scan-actions/gcp-python-teams-notification/README.md:14]:
```suggestion
1. Open in Cloud Shell.
```

File: post-scan-actions/gcp-python-teams-notification/README.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+[![Open in Cloud Shell](https://gstatic.com/cloudssh/images/open-btn.svg)](https://shell.cloud.google.com/cloudshell/editor?cloudshell_git_repo=https%3A%2F%2Fgithub.com%2Ftrendmicro%2Fcloudone-filestorage-plugins.git&cloudshell_workspace=post-scan-actions%2Fgcp-python-teams-notification&cloudshell_tutorial=docs/deploy-tutorial.md)
+
+### Local Machine
+
+1. Login to the Google Cloud SDK
```
Comment (davidblanchard123) [inline post-scan-actions/gcp-python-teams-notification/README.md:20]:
```suggestion
1. Login to the Google Cloud SDK.
```

File: post-scan-actions/gcp-python-teams-notification/README.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+   ```
+   gcloud config set project <PROJECT_ID>
+   ```
+
+## Configure Google Cloud function
+
+1. Specify the following fields and execute the deployment script:
+
+- **TEAMS_URL** - The incoming webhook URL generated from MS Teams Channel connectors. This generated URL can be created by following the step-by-step guide to creating an Incoming Webhook described here - [How to configure and use Incoming Webhooks in Microsoft Teams](https://techcommunity.microsoft.com/t5/microsoft-365-pnp-blog/how-to-configure-and-use-incoming-webhooks-in-microsoft-teams/ba-p/2051118).
```
Comment (davidblanchard123) [inline post-scan-actions/gcp-python-teams-notification/README.md:42]:
```suggestion
- **TEAMS_URL** - The incoming webhook URL generated from MS Teams Channel connectors. This generated URL can be created by following the step-by-step guide found here - [How to configure and use Incoming Webhooks in Microsoft Teams](https://techcommunity.microsoft.com/t5/microsoft-365-pnp-blog/how-to-configure-and-use-incoming-webhooks-in-microsoft-teams/ba-p/2051118).
```

File: post-scan-actions/gcp-python-teams-notification/README.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+- **DEPLOYMENT_REGION** - The region where the File Storage Security Storage stack was deployed.
```
Comment (davidblanchard123) [inline post-scan-actions/gcp-python-teams-notification/README.md:43]:
```suggestion
- **DEPLOYMENT_REGION** - The region in which the File Storage Security Storage stack is deployed.
```

File: post-scan-actions/gcp-python-teams-notification/README.md [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+## Deploy Google Cloud function to push Slack notifications
+
+1. Install Serverless on your local machine.
+
+   ```sh
+   npm install -g serverless
+   ```
+
+2. Deploy Serverless project.
```
Comment (davidblanchard123) [inline post-scan-actions/gcp-python-teams-notification/README.md:56]:
```suggestion
2. Deploy the Serverless project:
```

File: post-scan-actions/gcp-python-teams-notification/README.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+   --param="TEAMS_URL=<TEAMS_URL>" /
+   --param="DEPLOYMENT_REGION=<DEPLOYMENT_REGION>" /
+   --param="GCP_PROJECT_ID=<GCP_PROJECT_ID>" /
+   --param="TRIGGER_RESOURCE=<TRIGGER_RESOURCE>" /
+   --param="EVENT_TYPE=<EVENT_TYPE>"
+   ```
+
+## Test MS Teams notifications
+
+Check MS Teams to see new notifications. To test your deployment, you'll need to generate a malware detection using the eicar file.
```
Comment (davidblanchard123) [inline post-scan-actions/gcp-python-teams-notification/README.md:71]:
```suggestion
Check MS Teams for new notifications. To test your deployment, you'll need to generate a malware detection using the eicar file.
```

File: post-scan-actions/gcp-python-teams-notification/README.md [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+    ```
+    wget https://secure.eicar.org/eicar.com.txt
+
+    gsutil cp eicar.com.txt gs://<SCANNING_BUCKET_NAME>/eicar
+    ```
+
+   > File Storage Security scans the file and detects the malware.
+
+2. Execute the script to examine the scan result:
```
Comment (davidblanchard123) [inline post-scan-actions/gcp-python-teams-notification/README.md:83]:
```suggestion
2. Execute the script to examine the scan results:
```

File: post-scan-actions/gcp-python-teams-notification/docs/deploy-tutorial.md [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+--------------------------------
+
+## Configure Google Cloud function
+
+Build your deployment command by substituting the following fields in your sample deployment command. An example/default value for these fields can be found in the field description below.
+
+### Deployment by command-line
+
+Replace the values in the following command and store the output for the next step
```
Comment (davidblanchard123) [inline post-scan-actions/gcp-python-teams-notification/docs/deploy-tutorial.md:31]:
```suggestion
Replace the values in the following command and store the output for the next step:
```

File: post-scan-actions/gcp-python-teams-notification/docs/deploy-tutorial.md [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+    --param="TEAMS_URL=<TEAMS_URL>" \
+    --param="DEPLOYMENT_REGION=<DEPLOYMENT_REGION>" \
+    --param="GCP_PROJECT_ID=<GCP_PROJECT_ID>" \
+    --param="TRIGGER_RESOURCE=<TRIGGER_RESOURCE>" \
+    --param="EVENT_TYPE=<EVENT_TYPE>"
+```
+
+where,
+
+- **TEAMS_URL** - The incoming webhook URL generated from MS Teams Channel connectors. This generated URL can be created by following the step-by-step guide to creating an Incoming Webhook described here - [How to configure and use Incoming Webhooks in Microsoft Teams](https://techcommunity.microsoft.com/t5/microsoft-365-pnp-blog/how-to-configure-and-use-incoming-webhooks-in-microsoft-teams/ba-p/2051118).
```
Comment (davidblanchard123) [inline post-scan-actions/gcp-python-teams-notification/docs/deploy-tutorial.md:44]:
```suggestion
- **TEAMS_URL** - The incoming webhook URL generated from MS Teams Channel connectors. This generated URL can be created by following the step-by-step guide found here - [How to configure and use Incoming Webhooks in Microsoft Teams](https://techcommunity.microsoft.com/t5/microsoft-365-pnp-blog/how-to-configure-and-use-incoming-webhooks-in-microsoft-teams/ba-p/2051118).
```

File: post-scan-actions/gcp-python-teams-notification/docs/deploy-tutorial.md [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+- **DEPLOYMENT_REGION** - The region where the File Storage Security Storage stack was deployed.
```
Comment (davidblanchard123) [inline post-scan-actions/gcp-python-teams-notification/docs/deploy-tutorial.md:45]:
```suggestion
- **DEPLOYMENT_REGION** - The region in which the File Storage Security Storage stack is deployed.
```

File: post-scan-actions/gcp-python-teams-notification/docs/deploy-tutorial.md [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+- **GCP_PROJECT_ID** - Project ID of the GCP project.
+- **TRIGGER_RESOURCE** - Topic name of the scan result topic name. Example: `projects/<PROJECT_ID>/topics/<SCAN_RESULT_TOPIC_NAME>`
+- **EVENT_TYPE** - Optional. Defaults to `providers/cloud.pubsub/eventTypes/topic.publish`
+
+### Deployment through serverless.yml file
+
+You could hardcode these values in the <walkthrough-editor-select-line filePath="cloudone-filestorage-plugins/post-scan-actions/gcp-python-teams-notification/serverless.yml" startLine="6" endLine="12">serverless.yml</walkthrough-editor-select-line> file. To override a hard-coded value during runtime, simply pass it as a `--param` as shown in the command-line section above.
```
Comment (davidblanchard123) [inline post-scan-actions/gcp-python-teams-notification/docs/deploy-tutorial.md:52]:
```suggestion
You could hardcode these values in the <walkthrough-editor-select-line filePath="cloudone-filestorage-plugins/post-scan-actions/gcp-python-teams-notification/serverless.yml" startLine="6" endLine="12">serverless.yml</walkthrough-editor-select-line> file. To override a hard-coded value during runtime, pass it as a `--param` as shown in the command-line section above.
```

File: post-scan-actions/gcp-python-teams-notification/docs/deploy-tutorial.md [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+
+Simply replace, the `params` section of the serverless.yml with the right values and run `serverless deploy` as shown in the next step.
```
Comment (davidblanchard123) [inline post-scan-actions/gcp-python-teams-notification/docs/deploy-tutorial.md:54]:
```suggestion
Replace the `params` section of the serverless.yml with the correct values and run `serverless deploy` as shown in the next step.
```

File: post-scan-actions/gcp-python-teams-notification/docs/deploy-tutorial.md [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+
+1. Install Serverless on your local machine.
+
+    ```sh
+    npm install -g serverless
+    ```
+
+2. Deploy Serverless project.
+
+    This cloudone plugin requires some dependencies from the serverless plugin ecosystem, like `serverless-google-cloudfunctions` before we can deploy this project.
```
Comment (davidblanchard123) [inline post-scan-actions/gcp-python-teams-notification/docs/deploy-tutorial.md:78]:
```suggestion
    This cloudone plugin requires some dependencies from the serverless plugin ecosystem, for example `serverless-google-cloudfunctions`, before you can deploy this project.
```

File: post-scan-actions/gcp-python-teams-notification/docs/deploy-tutorial.md [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+
+    - Install serverless dependencies
```
Comment (davidblanchard123) [inline post-scan-actions/gcp-python-teams-notification/docs/deploy-tutorial.md:80]:
```suggestion
    - Install the serverless dependencies:
```

File: post-scan-actions/gcp-python-teams-notification/docs/deploy-tutorial.md [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+
+        ```sh
+        serverless plugin install -n serverless-google-cloudfunctions
+        ```
+
+    - Deploy your serverless project
```
Comment (davidblanchard123) [inline post-scan-actions/gcp-python-teams-notification/docs/deploy-tutorial.md:86]:
```suggestion
    - Deploy your serverless project:
```

File: post-scan-actions/gcp-python-teams-notification/docs/deploy-tutorial.md [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+
+        - Through command-line
```
Comment (davidblanchard123) [inline post-scan-actions/gcp-python-teams-notification/docs/deploy-tutorial.md:88]:
```suggestion
        - Through the command-line:
```

File: post-scan-actions/gcp-python-teams-notification/docs/deploy-tutorial.md [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+            ```sh
+            serverless deploy -s prod \
+            --param="TEAMS_URL=<TEAMS_URL>" \
+            --param="DEPLOYMENT_REGION=<DEPLOYMENT_REGION>" \
+            --param="GCP_PROJECT_ID=<GCP_PROJECT_ID>" \
+            --param="TRIGGER_RESOURCE=<TRIGGER_RESOURCE>" \
+            --param="EVENT_TYPE=<EVENT_TYPE>"
+            ```
+
+        - Using `serverless.yml` and the values provided within the file
```
Comment (davidblanchard123) [inline post-scan-actions/gcp-python-teams-notification/docs/deploy-tutorial.md:99]:
```suggestion
        - Using `serverless.yml` and the values provided within the file:
```

File: post-scan-actions/gcp-python-teams-notification/docs/deploy-tutorial.md [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+
+            ```sh
+            serverless deploy -s prod
+            ```
+
+--------------------------------
+
+## Test MS Teams notifications
+
+Check MS Teams to see new notifications. To test your deployment, you'll need to generate a malware detection using the eicar file.
```
Comment (davidblanchard123) [inline post-scan-actions/gcp-python-teams-notification/docs/deploy-tutorial.md:109]:
```suggestion
Check MS Teams for new notifications. To test your deployment, you'll need to generate a malware detection using the eicar file.
```

File: post-scan-actions/gcp-python-teams-notification/docs/deploy-tutorial.md [hunk 21]
```diff
@@ -0,0 +1,138 @@
...
+
+    ```
+    wget https://secure.eicar.org/eicar.com.txt
+
+    gsutil cp eicar.com.txt gs://<SCANNING_BUCKET_NAME>/eicar
+    ```
+
+    > File Storage Security scans the file and detects the malware.
+
+2. Execute the script to examine the scan result:
```
Comment (davidblanchard123) [inline post-scan-actions/gcp-python-teams-notification/docs/deploy-tutorial.md:121]:
```suggestion
2. Execute the script to examine the scan results:
```

**Review by davidblanchard123 — COMMENTED** [review]
Made some suggestions

**Review by jack-c-tang — APPROVED** [review]

**Pull request merged**